memory/src/main/scala/filodb.memory/data/OffheapLFSortedIDMap.scala

```diff
@@ -55,8 +72,82 @@ object OffheapLFSortedIDMap extends StrictLogging {
 
   val IllegalStateResult = -1  // Returned from binarySearch when state changed underneath
 
+  val InitialExclusiveRetryTimeoutNanos = 1.millisecond.toNanos
+  val MaxExclusiveRetryTimeoutNanos = 1.second.toNanos
+
   val _logger = logger
 
+  val sharedLockLingering = Kamon.counter("memory-shared-lock-lingering")
+
+  // Tracks all the shared locks held, by each thread.
+  val sharedLockCounts = new ThreadLocal[Map[MapHolder, Int]]
+
+  // Lock state memory offsets for all known MapHolder classes.
+  val lockStateOffsets = new ConcurrentHashMap[Class[_ <: MapHolder], Long]
+
+  // Updates the shared lock count, for the current thread.
+  //scalastyle:off
+  def adjustSharedLockCount(inst: MapHolder, amt: Int): Unit = {
```

Comment:
I'm worried about this method and the next one, as well as the accompanying data structures.   Updating these structures are not cheap, and they could potentially use a ton of memory too (every Long and Int will be boxed, in both data structures).  When will they be updated?  When will the methods be called?  Do they grow based on the number of locks held?

Reply:
A few extra object allocations is a bit small compared to all the others created during a query. I don't know how Scala int boxing works, but in Java, small values are cached.